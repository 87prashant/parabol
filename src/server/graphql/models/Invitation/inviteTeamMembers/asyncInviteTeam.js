import getRethink from 'server/database/rethinkDriver';
import {getUserId} from 'server/utils/authorization';
import getInviterInfoAndTeamName from './getInviterInfoAndTeamName';
import createEmailPromises from './createEmailPromises';
import resolveSentEmails from './resolveSentEmails';
import makeInviteToken from './makeInviteToken';
import makeInvitationsForDB from './makeInvitationsForDB';

export default async function asyncInviteTeam (authToken, teamId, invitees) {
  const r = getRethink();
  const userId = getUserId(authToken);
  const inviteesWithTokens = invitees.map(invitee => ({...invitee, inviteToken: makeInviteToken()}));
  const inviterInfoAndTeamName = await getInviterInfoAndTeamName(teamId, userId);
  const sendEmailPromises = createEmailPromises(inviterInfoAndTeamName, inviteesWithTokens);
  const {inviteesToStore} = await resolveSentEmails(sendEmailPromises, inviteesWithTokens);
  const invitationsForDB = await makeInvitationsForDB(inviteesToStore, teamId, userId);
  // Bulk insert, wait in case something queries the invitation table
  await r.table('Invitation').insert(invitationsForDB);
  return true;
  // TODO generate email to inviter including folks that we couldn't reach
  // if (inviteeErrors.length > 0) {
  // throw errorObj({_error: 'Some invitations were not sent', type: 'inviteSendFail', failedEmails: inviteeErrors});
  // }
};
