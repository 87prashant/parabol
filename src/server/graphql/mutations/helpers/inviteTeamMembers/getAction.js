import {ASK_APPROVAL, REACTIVATE, SEND_INVITATION} from 'server/utils/serverConstants';
import {ALREADY_ON_TEAM, PENDING_APPROVAL, SUCCESS} from 'universal/utils/constants';

const getAction = (invitee, inviterIsBillingLeader) => {
  const {
    isActiveTeamMember,
    isPendingApproval,
    isPendingInvitation,
    isNewTeamMember,
    isOrgMember
  } = invitee;
  if (isActiveTeamMember) return ALREADY_ON_TEAM;
  if (isPendingApproval && !inviterIsBillingLeader) return PENDING_APPROVAL;
  if (isPendingInvitation) return SUCCESS;
  if (isOrgMember) {
    return isNewTeamMember ? SEND_INVITATION : REACTIVATE;
  }
  return inviterIsBillingLeader ? SEND_INVITATION : ASK_APPROVAL;
};

export default getAction;
