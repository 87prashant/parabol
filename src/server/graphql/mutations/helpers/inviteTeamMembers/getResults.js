import {ASK_APPROVAL, REACTIVATE, SEND_INVITATION} from 'server/utils/serverConstants';
import {PENDING_APPROVAL, REACTIVATED, SUCCESS} from 'universal/utils/constants';

const resultLookup = {
  [ASK_APPROVAL]: PENDING_APPROVAL,
  [SEND_INVITATION]: SUCCESS,
  [REACTIVATE]: REACTIVATED
};

const getResults = (detailedInvitations) => {
  return detailedInvitations.map((invitee) => {
    const {email, action, preferredName} = invitee;
    return {
      email,
      result: resultLookup[action],
      preferredName
    };
  });
};

export default getResults;
