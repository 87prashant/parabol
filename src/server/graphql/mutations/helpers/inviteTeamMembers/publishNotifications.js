import getPubSub from 'server/utils/getPubSub';
import {NOTIFICATIONS_ADDED, NOTIFICATIONS_CLEARED} from 'universal/utils/constants';
import mergeObjectsWithArrValues from 'universal/utils/mergeObjectsWithArrValues';

const publishNotifications = (notificationsToSend) => {
  const {
    reactivation,
    notificationsToClear,
    teamInvite,
    approval
  } = notificationsToSend;

  const allNotificationsToAdd = mergeObjectsWithArrValues(reactivation, teamInvite, approval);
  Object.keys(allNotificationsToAdd).forEach((userId) => {
    const notificationsAdded = {
      notifications: allNotificationsToAdd[userId]
    };
    getPubSub().publish(`${NOTIFICATIONS_ADDED}.${userId}`, {notificationsAdded});
  });
  Object.keys(notificationsToClear).forEach((userId) => {
    const notificationsCleared = notificationsToClear[userId];
    getPubSub().publish(`${NOTIFICATIONS_CLEARED}.${userId}`, {notificationsCleared});
  });
};

export default publishNotifications;
