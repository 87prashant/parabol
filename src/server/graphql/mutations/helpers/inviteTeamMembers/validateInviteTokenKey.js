import bcrypt from 'bcryptjs'
import promisify from 'es6-promisify'

const compare = promisify(bcrypt.compare)

export default function validateInviteTokenKey (key, hashStringToCompare) {
  return compare(key, hashStringToCompare)
}
