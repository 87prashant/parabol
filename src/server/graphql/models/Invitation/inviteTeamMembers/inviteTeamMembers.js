import {GraphQLBoolean, GraphQLID, GraphQLList, GraphQLNonNull} from 'graphql';
import getRethink from 'server/database/rethinkDriver';
import asyncInviteTeam from 'server/graphql/models/Invitation/inviteTeamMembers/asyncInviteTeam';
import getInviterInfoAndTeamName from 'server/graphql/models/Invitation/inviteTeamMembers/getInviterInfoAndTeamName';
import inviteAsUser from 'server/graphql/models/Invitation/inviteTeamMembers/inviteAsUser';
import {getUserId, getUserOrgDoc, isBillingLeader, requireOrgLeaderOrTeamMember} from 'server/utils/authorization';
import getPubSub from 'server/utils/getPubSub';
import tmsSignToken from 'server/utils/tmsSignToken';
import shortid from 'shortid';
import {ADD_TO_TEAM, PRESENCE, REJOIN_TEAM, USER_MEMO} from 'universal/subscriptions/constants';
import {ALREADY_ON_TEAM, PENDING_APPROVAL, REACTIVATED, SUCCESS, TEAM_INVITE} from 'universal/utils/constants';
import {Invitee} from '../invitationSchema';


//const determineInviteeAction = (invitee) => {
//  const {
//    isActiveTeamMember,
//    isPendingApproval,
//    isPendingInvitation,
//    isUser,
//    isActiveUser,
//    isOrgMember,
//    isNewTeamMember,
//  } = invitee;
//  if (isActiveTeamMember) {
//    return
//  }
//}

const SEND_NOTIFICATION = 'SEND_NOTIFICATION';
const SEND_EMAIL = 'SEND_EMAIL';
const ASK_APPROVAL = 'ASK_APPROVAL';
//const ASK_APPROVAL_EMAIL = 'ASK_APPROVAL_EMAIL';
//const ASK_APPROVAL_EMAIL_NOTIFICATION = 'ASK_APPROVAL_EMAIL_NOTIFICATION';
//const ASK_APPROVAL_NOTIFICATION = 'ASK_APPROVAL_NOTIFICATION';

const sendNotification = async (invitees, inviter) => {
  if (invitees.length === 0) return invitees;
  const r = getRethink();
  const now = new Date();
  const {orgId, inviterName, teamId, teamName} = inviter;
  const invitations = invitees.map((invitee) => ({
    id: shortid.generate(),
    type: TEAM_INVITE,
    startAt: now,
    orgId,
    userIds: [invitee.userId],
    varList: [inviterName, teamId, teamName]
  }));
  await r.table('Notification').insert(invitations);
  return true;
};

const reactivateAndSendWelcomeBack = async (invitees, inviter, exchange) => {
  if (invitees.length === 0) return invitees;
  const {teamId, teamName} = inviter;
  const r = getRethink();
  const userIds = invitees.map(({userId}) => userId);
  const teamMemberIds = userIds.map((userId) => `${userId}::${teamId}`);
  const reactivatedUsers = await r.table('TeamMember')
    .getAll(r.args(teamMemberIds), {index: 'id'})
    .update({isNotRemoved: true})
    .do(() => {
      return r.table('User')
        .getAll(r.args(userIds))
        .update((user) => {
          return user.merge({
            tms: user('tms').append(teamId)
          });
        }, {returnChanges: true})('changes')('new_val')
        .default(null);
    });
  reactivatedUsers.forEach((user) => {
    const {preferredName, id: reactivatedUserId, tms} = user;
    getPubSub().publish(`${USER_MEMO}/${reactivatedUserId}`, {
      type: ADD_TO_TEAM,
      teamId,
      teamName,
      _authToken: tmsSignToken({sub: reactivatedUserId}, tms)
    });
    const channel = `${PRESENCE}/${teamId}`;
    // TODO refactor out exchange
    exchange.publish(channel, {
      type: REJOIN_TEAM,
      name: preferredName,
      sender: inviter.userId
    });
  });
  return true;
}

//const handleInOrgInvite = async (invitee, inviter) => {
//  const {isNewTeamMember, email, userId} = invitee;
//  if (isNewTeamMember) {
//    await sendNotification(userId, inviter);
//    return {
//      result: SUCCESS,
//      email
//    };
//  } else {
//
//    reactivateAndSendWelcomeBack();
//  }
//};

const getAction = (invitee, inviterIsBillingLeader) => {
  const {
    isActiveTeamMember,
    isPendingApproval,
    isPendingInvitation,
    isActiveUser,
    isNewTeamMember,
    isOrgMember,
  } = invitee;
  if (isActiveTeamMember) return ALREADY_ON_TEAM;
  if (isPendingApproval && !inviterIsBillingLeader) return PENDING_APPROVAL;
  if (isPendingInvitation) return SUCCESS;
  if (isOrgMember) {
    return isNewTeamMember ? SEND_NOTIFICATION : REACTIVATED;
  }
  if (inviterIsBillingLeader) {
    return isActiveUser ? SEND_NOTIFICATION : SEND_EMAIL;
  }
  return ASK_APPROVAL;
};

//const getInvitationErrors = (inviteeEmails, activeTeamMembers = [], pendingInvitations = [], pendingApprovals = []) => {
//  const getInviteeError = (inviteeEmail) => {
//    if (activeTeamMembers.includes(inviteeEmail)) {
//      return 'member'
//    }
//    if (pendingInvitations.includes(inviteeEmail)) {
//      return 'invited'
//    }
//    if (pendingApprovals.includes(inviteeEmail)) {
//      return 'pending'
//    }
//    return undefined;
//  };
//
//  return inviteeEmails.map((inviteeEmail) => ({
//    email: inviteeEmail,
//    error: getInviteeError(inviteeEmail)
//  }));
//};

export default {
  type: GraphQLBoolean,
  description: `If in the org,
     Send invitation emails to a list of email addresses, add them to the invitation table.
     Else, send a request to the org leader to get them approval and put them in the OrgApproval table.
     Returns true if invitation is sent, false if approval is needed, undefined/null if neither`,
  args: {
    teamId: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'The id of the inviting team'
    },
    invitees: {
      type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(Invitee)))
    }
    // notificationId: {
    //   type: GraphQLID
    // }
  },
  async resolve(source, {invitees, teamId}, {authToken, exchange}) {
    const r = getRethink();
    const now = Date.now();

    // AUTH
    await requireOrgLeaderOrTeamMember(authToken, teamId);
    const userId = getUserId(authToken);
    const {name: teamName, orgId} = await r.table('Team').get(teamId).pluck('name', 'orgId');
    const userOrgDoc = await getUserOrgDoc(userId, orgId);
    const inviterIsBillingLeader = isBillingLeader(userOrgDoc);

    // VALIDATION
    // don't let them invite the same person twice
    const emailArr = invitees.map((invitee) => invitee.email);
    const {pendingInvitations, pendingApprovals, teamMembers, users} = await r.expr({
      pendingInvitations: r.table('Invitation')
        .getAll(r.args(emailArr), {index: 'email'})
        .filter((invitation) => invitation('tokenExpiration').ge(r.epochTime(now)))('email')
        .coerceTo('array'),
      pendingApprovals: r.table('OrgApproval')
        .getAll(r.args(emailArr), {index: 'email'})
        .filter({teamId})
        .coerceTo('array'),
      teamMembers: r.table('TeamMember')
        .getAll(teamId, {index: 'teamId'})
        //.merge((teamMember) => ({
        //  userOrgs: r.table('User').get(teamMember('userId'))('userOrgs').default([])
        //}))
        .coerceTo('array'),
      users: r.table('User')
        .getAll(r.args(emailArr), {index: 'email'})
        .coerceTo('array')
    });

    //const newInvitations = getInvitationErrors(emailArr, activeTeamMembers, pendingInvitations, pendingApprovals);

    // RESOLUTION
    const activeTeamMembers = teamMembers.filter((m) => m.isNotRemoved === true).map((m) => m.email);
    const inactiveTeamMembers = teamMembers.filter((m) => !m.isNotRemoved).map((m) => m.email);
    const inviterAndTeamName = await getInviterInfoAndTeamName(teamId, userId);
    const inviterDetails = {
      ...inviterAndTeamName,
      orgId,
      teamId,
      isBillingLeader: inviterIsBillingLeader
    };

    const detailedInvitations = emailArr.map((email) => {
      const userDoc = users.find((user) => user.email === email);
      const details = {
        email,
        isActiveTeamMember: activeTeamMembers.includes(email),
        isPendingApproval: pendingApprovals.includes(email),
        isPendingInvitation: pendingInvitations.includes(email),
        isUser: Boolean(userDoc),
        isActiveUser: userDoc && !userDoc.inactive,
        isOrgMember: userDoc && userDoc.userOrgs.includes((userDocOrg) => userDocOrg.id === orgId),
        isNewTeamMember: !inactiveTeamMembers.includes((tm) => tm.email === email),
        userId: userDoc && userDoc.id
      };
      return {
        ...details,
        action: getAction(details, inviterIsBillingLeader)
      }
    });

    const inviteesToNotify = detailedInvitations.filter(({action}) => action === SEND_NOTIFICATION);
    const inviteesToReactivate = detailedInvitations.filter(({action}) => action === REACTIVATED);
    const inviteesToEmail = detailedInvitations.filter(({action}) => action === SEND_EMAIL);
    const approveWithNotification = detailedInvitations.filter(({action}) => action === ASK_APPROVAL_NOTIFICATION || action === ASK_APPROVAL_EMAIL_NOTIFICATION);
    const approveWithEmail = detailedInvitations.filter(({action}) => action === ASK_APPROVAL_EMAIL || action === ASK_APPROVAL_EMAIL_NOTIFICATION);

    const promises = [
      sendNotification(inviteesToNotify, inviterDetails),
      reactivateAndSendWelcomeBack(inviteesToReactivate, inviterDetails),
      asyncInviteTeam(userId, teamId, inviteesToEmail)


    ]

    const results = await Promise.all(detailedInvitations.map((invitee) => {
      const {
        isActiveTeamMember,
        isPendingApproval,
        isPendingInvitation,
        isUser,
        isActiveUser,
        isOrgMember,
        isNewTeamMember
      } = invitee;

    }));


    //const idsToReactivate = [];
    //const newInvitees = [];
    //for (let i = 0; i < validInvitees.length; i++) {
    //  const validInvitee = validInvitees[i];
    //  const inactiveInvitee = inactiveTeamMembers.find((m) => m.email === validInvitee.email);
    //  if (inactiveInvitee) {
    //    // if they were previously removed from the team, see if they're still in the org
    //    const inOrg = Boolean(inactiveInvitee.userOrgs.find((userOrg) => userOrg.id === orgId));
    //    if (inOrg) {
    //      // if they're in the org, reactive them
    //      idsToReactivate.push(inactiveInvitee.id);
    //      continue;
    //    }
    //  }
    //  newInvitees.push(validInvitee);
    //}
    //
    //await reactivateTeamMembers(idsToReactivate, teamId, teamName, exchange, userId);
    //
    //if (newInvitees.length > 0) {
    //  // if it's a Billing Leader send them all
    //  const inviteeEmails = newInvitees.map((i) => i.email);
    //  if (inviterIsBillingLeader) {
    //    // if any folks were pending, release the floodgates, a Billing Leader has approved them
    //    const pendingApprovals = await removeOrgApprovalAndNotification(orgId, inviteeEmails);
    //    // we should always have a fresh invitee, but we do this safety check in case the front-end validation messes up
    //    const freshInvitees = newInvitees.filter((i) =>
    //      !pendingApprovals.find((d) => d.inviteeEmail === i.email && d.invitedTeamId === teamId));
    //    if (freshInvitees) {
    //      setTimeout(async () => {
    //        await asyncInviteTeam(userId, teamId, freshInvitees);
    //      }, 0);
    //    }
    //    pendingApprovals.forEach((invite) => {
    //      const {inviterId, inviteeEmail, invitedTeamId} = invite;
    //      const invitee = [{email: inviteeEmail}];
    //      // when we invite the person, try to invite from the original requester, if not, Billing Leader
    //      setTimeout(async () => {
    //        await asyncInviteTeam(inviterId, invitedTeamId, invitee);
    //      }, 0);
    //    });
    //
    //    return true;
    //  }
    //  // return false if org approvals sent, true if only invites were sent
    return inviteAsUser(newInvitees, orgId, userId, teamId, teamName);
    //}
    return undefined;
  }
};

