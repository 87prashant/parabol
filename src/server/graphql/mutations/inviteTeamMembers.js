import {GraphQLID, GraphQLList, GraphQLNonNull} from 'graphql';
import Invitee from 'server/graphql/types/Invitee';
import InviteTeamMembersPayload from 'server/graphql/types/InviteTeamMembersPayload';
import inviteTeamMembers from 'server/safeMutations/inviteTeamMembers';
import {getUserId, requireOrgLeaderOrTeamMember} from 'server/utils/authorization';
import publish from 'server/utils/publish';
import {TEAM} from 'universal/subscriptions/constants';
import {ADDED, NOTIFICATION, ORG_APPROVAL, REJOIN_TEAM, REMOVED, TEAM_MEMBER} from 'universal/utils/constants';
import fromTeamMemberId from 'universal/utils/relay/fromTeamMemberId';

export default {
  type: new GraphQLNonNull(InviteTeamMembersPayload),
  description: `If in the org,
     Send invitation emails to a list of email addresses, add them to the invitation table.
     Else, send a request to the org leader to get them approval and put them in the OrgApproval table.`,
  args: {
    teamId: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'The id of the inviting team'
    },
    invitees: {
      type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(Invitee)))
    }
  },
  async resolve(source, {invitees, teamId}, {authToken, dataLoader, socketId: mutatorId}) {
    const operationId = dataLoader.share();
    // AUTH
    await requireOrgLeaderOrTeamMember(authToken, teamId);
    const viewerId = getUserId(authToken);

    // RESOLUTION
    const subOptions = {mutatorId, operationId};
    const {
      orgApprovalIds,
      reactivations,
      results,
      removedOrgApprovals,
      removedRequestNotifications,
      requestNotifications
    } = await inviteTeamMembers(invitees, teamId, viewerId, subOptions);
    const reactivatedTeamMemberIds = reactivations.map(({teamMemberId}) => teamMemberId);

    // HANDLE REACTIVATION
    reactivations.forEach(({notificationId, teamMemberId, preferredName}) => {
      // send a team member + temporary toast to the rest of the users
      const notification = {type: REJOIN_TEAM, teamId, preferredName};
      publish(TEAM_MEMBER, teamId, ADDED, {teamId, notification}, subOptions);
      // send a team + persisted notification to the reactivated team member
      const {userId} = fromTeamMemberId(teamMemberId);
      publish(TEAM, userId, ADDED, {teamId, notificationId}, subOptions);
    });

    // HANDLE NON-ORG LEADERS NEEDING APPROVAL
    orgApprovalIds.forEach((orgApprovalId) => {
      publish(ORG_APPROVAL, teamId, ADDED, {orgApprovalId}, subOptions);
    });
    requestNotifications.forEach((requestNotification) => {
      const {id: notificationId, userIds} = requestNotification;
      userIds.forEach((userId) => {
        // if the org leader triggers this it won't be in the mutation payload, so mutatorId is not passed in
        publish(NOTIFICATION, userId, ADDED, {notificationId}, {operationId});
      });
    });

    // HANDLE ORG LEADER INVITING A PENDING ORG APPROVAL INVITEE
    const removedOrgApprovalIds = removedOrgApprovals.map(({id}) => id);
    removedOrgApprovalIds.forEach((removedOrgApprovalId) => {
      publish(ORG_APPROVAL, teamId, REMOVED, {removedOrgApprovalId}, subOptions);
    });
    removedRequestNotifications.forEach((notification) => {
      const {userIds} = notification;
      userIds.forEach((userId) => {
        // if the org leader triggers this it won't be in the mutation payload, so mutatorId is not passed in
        publish(NOTIFICATION, userId, REMOVED, {notification}, {operationId});
      });
    });
    return {orgApprovalIds, reactivatedTeamMemberIds, removedOrgApprovalIds, results};
  }
};

