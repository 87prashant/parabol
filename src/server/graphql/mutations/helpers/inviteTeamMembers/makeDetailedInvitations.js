import getAction from 'server/graphql/mutations/helpers/inviteTeamMembers/getAction';
import {APPROVED, PENDING} from 'server/utils/serverConstants';

const makeDetailedInvitations = (
  teamMembers,
  emailArr = [],
  users = [],
  orgApprovals = [],
  pendingInvitations = [],
  inviter = {}
) => {
  const {orgId, isBillingLeader, teamId} = inviter;
  const pendingTeamApprovals = orgApprovals.filter(
    (approval) => approval.teamId === teamId && approval.status === PENDING
  );
  const approvedOrgApprovals = orgApprovals.filter((approval) => approval.status === APPROVED);
  return emailArr.map((email) => {
    const userDoc = users.find((user) => user.email === email);
    const teamMemberDoc = teamMembers.find((m) => m.email === email);
    const details = {
      email,
      isActiveTeamMember: teamMemberDoc && teamMemberDoc.isNotRemoved,
      isPendingApproval: Boolean(pendingTeamApprovals.find((approval) => approval.email === email)),
      isPendingInvitation: pendingInvitations.includes(email),
      isUser: Boolean(userDoc),
      isOrgMember:
        userDoc && Boolean(userDoc.userOrgs.find((userDocOrg) => userDocOrg.id === orgId)),
      isNewTeamMember: !teamMemberDoc,
      teamMemberId: teamMemberDoc && teamMemberDoc.id,
      userId: userDoc && userDoc.id,
      preferredName: userDoc && userDoc.preferredName,
      isApproved:
        Boolean(approvedOrgApprovals.find((approval) => approval.email === email)) ||
        isBillingLeader ||
        false
    };
    return {
      ...details,
      action: getAction(details)
    };
  });
};

export default makeDetailedInvitations;
