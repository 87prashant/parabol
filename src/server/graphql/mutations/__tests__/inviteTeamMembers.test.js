import MockDate from 'mockdate';
import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
import MockDB from 'server/__tests__/setup/MockDB';
import {__now} from 'server/__tests__/setup/mockTimes';
import expectAsyncToThrow from 'server/__tests__/utils/expectAsyncToThrow';
import * as publishNotifications from 'server/graphql/mutations/helpers/inviteTeamMembers/publishNotifications';
import inviteTeamMembers from 'server/graphql/mutations/inviteTeamMembers';
import * as asyncInviteTeam from 'server/safeMutations/asyncInviteTeam';
import * as createPendingApprovals from 'server/safeMutations/createPendingApprovals';
import * as reactivateTeamMembersAndMakeNotifications from 'server/safeMutations/reactivateTeamMembersAndMakeNotifications';
import * as removeOrgApprovalAndNotification from 'server/safeMutations/removeOrgApprovalAndNotification';
import * as sendInvitationViaNotification from 'server/safeMutations/sendInvitationViaNotification';
import {REQUEST_NEW_USER} from 'universal/utils/constants';
import newInvitee from 'server/__tests__/utils/newInvitee';
import DynamicSerializer from 'dynamic-serializer';

MockDate.set(__now);
console.error = jest.fn();

describe('inviteTeamMembers', () => {
  test('throws in the caller is not on the team', async () => {
    // SETUP
    asyncInviteTeam.default = jest.fn();
    sendInvitationViaNotification.default = jest.fn();
    const authToken = mockAuthToken({id: 1, tms: ['fakeTeam'], lastSeenAt: new Date()});
    const invitees = [];
    const teamId = 'realTeam';

    // VERIFY
    await expectAsyncToThrow(
      inviteTeamMembers.resolve(undefined, {invitees, teamId}, {authToken}),
      [teamId]
    );
  });

  test('reactivates a team member that has been invited back to the team', async () => {
    // SETUP
    asyncInviteTeam.default = jest.fn(() => []);
    sendInvitationViaNotification.default = jest.fn(() => []);
    reactivateTeamMembersAndMakeNotifications.default = jest.fn(() => []);
    removeOrgApprovalAndNotification.default = jest.fn(() => []);
    createPendingApprovals.default = jest.fn(() => []);
    publishNotifications.default = jest.fn(() => []);

    const mockDB = new MockDB();
    await mockDB.init()
      .teamMember(4, {isNotRemoved: false});
    const firstUser = mockDB.db.user[0];
    const authToken = mockAuthToken(firstUser);
    const invitees = [{email: mockDB.db.user[4].email}];
    const teamId = mockDB.context.team.id;
    // TEST
    const res = await inviteTeamMembers.resolve(undefined, {invitees, teamId}, {authToken});

    // VERIFY
    expect(reactivateTeamMembersAndMakeNotifications.default).toHaveBeenCalledTimes(1);
    expect(res).toMatchSnapshot();
  });

  test('invites a new person via email', async () => {

    // SETUP
    asyncInviteTeam.default = jest.fn(() => []);
    sendInvitationViaNotification.default = jest.fn(() => []);
    reactivateTeamMembersAndMakeNotifications.default = jest.fn(() => []);
    removeOrgApprovalAndNotification.default = jest.fn(() => []);
    createPendingApprovals.default = jest.fn(() => []);
    publishNotifications.default = jest.fn(() => []);
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    await mockDB.init()
    const firstUser = mockDB.db.user[0];
    const authToken = mockAuthToken(firstUser);
    const invitee = newInvitee();
    const invitees = [{email: invitee.email}];
    const teamId = mockDB.context.team.id;
    // TEST
    const res = await inviteTeamMembers.resolve(undefined, {invitees, teamId}, {authToken});

    // VERIFY
    expect(asyncInviteTeam.default).toHaveBeenCalledTimes(1);
    const staticRes = dynamicSerializer.toStatic(res, ['results.email']);
    expect(staticRes).toMatchSnapshot();
  });

  test('invites a new person via notification', async () => {
    // SETUP
    asyncInviteTeam.default = jest.fn(() => []);
    sendInvitationViaNotification.default = jest.fn(() => []);
    reactivateTeamMembersAndMakeNotifications.default = jest.fn(() => []);
    removeOrgApprovalAndNotification.default = jest.fn(() => []);
    createPendingApprovals.default = jest.fn(() => []);
    publishNotifications.default = jest.fn(() => []);
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const invitee = newInvitee();
    await mockDB.init()
      .newOrg({name: 'Sad Sacks, Inc.'})
      .newTeam({name: 'The Lonely Ones'})
      .newUser({name: 'Leader of One', email: invitee.email})
      .newTeamMember({isLead: true});
    const firstUser = mockDB.db.user[0];
    const authToken = mockAuthToken(firstUser);
    const invitees = [{email: mockDB.context.user.email}];
    const teamId = mockDB.context.team.id;
    // TEST
    const res = await inviteTeamMembers.resolve(undefined, {invitees, teamId}, {authToken});

    // VERIFY
    expect(sendInvitationViaNotification.default).toHaveBeenCalledTimes(1);
    expect(removeOrgApprovalAndNotification.default).toHaveBeenCalledTimes(1);
    const staticRes = dynamicSerializer.toStatic(res, ['results.email']);
    expect(staticRes).toMatchSnapshot();
  });

  test('asks the billing leader for approval', async () => {
    // SETUP
    asyncInviteTeam.default = jest.fn(() => []);
    sendInvitationViaNotification.default = jest.fn(() => []);
    reactivateTeamMembersAndMakeNotifications.default = jest.fn(() => []);
    removeOrgApprovalAndNotification.default = jest.fn(() => []);
    createPendingApprovals.default = jest.fn(() => []);
    publishNotifications.default = jest.fn(() => []);
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const invitee = newInvitee();
    await mockDB.init();
    const secondUser = mockDB.db.user[1];
    const authToken = mockAuthToken(secondUser);
    const invitees = [{email: invitee.email}];
    const teamId = mockDB.context.team.id;

    // TEST
    const res = await inviteTeamMembers.resolve(undefined, {invitees, teamId}, {authToken});

    // VERIFY
    expect(createPendingApprovals.default).toHaveBeenCalledTimes(1);
    const staticRes = dynamicSerializer.toStatic(res, ['results.email']);
    expect(staticRes).toMatchSnapshot();
  });
});
