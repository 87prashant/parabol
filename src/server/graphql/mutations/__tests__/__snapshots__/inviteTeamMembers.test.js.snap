// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inviteTeamMembers asks the billing leader for approval 1`] = `
Object {
  "results": Array [
    Object {
      "email": 0,
      "preferredName": undefined,
      "result": "PENDING_APPROVAL",
    },
  ],
}
`;

exports[`inviteTeamMembers invites a new person via email 1`] = `
Object {
  "results": Array [
    Object {
      "email": 0,
      "preferredName": undefined,
      "result": "SUCCESS",
    },
  ],
}
`;

exports[`inviteTeamMembers invites a new person via notification 1`] = `
Object {
  "results": Array [
    Object {
      "email": 0,
      "preferredName": "Leader of One",
      "result": "ALREADY_ON_TEAM",
    },
  ],
}
`;

exports[`inviteTeamMembers reactivates a team member that has been invited back to the team 1`] = `
Object {
  "results": Array [
    Object {
      "email": "bruce.banner@avengers.org",
      "preferredName": "Hulk",
      "result": "SUCCESS",
    },
  ],
}
`;

exports[`inviteTeamMembers throws in the caller is not on the team 1`] = `"{\\"_error\\":\\"Unauthorized to view details for team 0 with token {\\\\\\"iss\\\\\\":\\\\\\"action-test\\\\\\",\\\\\\"sub\\\\\\":1,\\\\\\"aud\\\\\\":\\\\\\"test-audience\\\\\\",\\\\\\"exp\\\\\\":1485849600,\\\\\\"iat\\\\\\":1483257600,\\\\\\"tms\\\\\\":[\\\\\\"fakeTeam\\\\\\"]}\\"}"`;
